India WRIS: Water Quality and Rainfall dataset

SenthamizhanV opened the pull request:
_Creating a draft PR since the place names for water quality datasets are not resolved yet._

@pradh @ajaits 
Please find the data imports for India WRIS datasets - Ground Water quality, Surface Water quality, Rainfall data.

beets commented on scripts/india_wris/base.py:
```diff
@@ -0,0 +1,156 @@
+        self.util_names = util_names
+        self.module_dir = os.path.dirname(__file__)
+
+        self.solute_mcf = template_strings['solute_mcf']
```
Nit: add assertions that all the required properties of template_strings are available?

SenthamizhanV replied:
Added assert statements in the `base.py` file. Please check if it's fine and readable.

beets replied:
this looks good!

beets commented on scripts/india_wris/base.py:
```diff
@@ -0,0 +1,156 @@
+from india.geo.districts import IndiaDistrictsMapper
+
+
+class WaterQualityBase():
```
Please add module level documentation for this base class.

SenthamizhanV replied:
Added docstring for base class

beets commented on scripts/india_wris/base.py:
```diff
@@ -0,0 +1,156 @@
+            os.path.join(self.module_dir,
+                         'data/{}.csv'.format(self.util_names)))
+        # Dropping rows with all data missing
+        self.df.dropna(thresh=11, inplace=True)
```
How was the value 11 chosen? Can it be derived, rather than hard-coded?

Also, were the dropped values inspected, and what were the issues with those rows? For processing scripts like this, it's best to print some reports as it runs, such as number of dropped rows, etc.

SenthamizhanV replied:
There were rows in the csv in which only the place names and latlongs are present. All columns corresponding water quality properties were empty. So I checked the files and chose 11 as common value for all datasets, which is the `minimum number of non-empty columns + 1`.

Now I added a [helper method](https://github.com/datacommonsorg/data/blob/9e1ab51d6e1551ff631a7f65492e92af9f7d8c26/scripts/india_wris/base.py#L60) to the class to calculate the minimum number of non-empty columns and drop rows with empty columns equal to that value.

beets commented on scripts/india_wris/base.py:
```diff
@@ -0,0 +1,156 @@
+        # Mapping district names to LGD Codes
+        mapper = IndiaDistrictsMapper()
+        df_map = self.df[['StateName',
+                          'DistrictName']].drop_duplicates().dropna()
```
ditto here, regarding inspecting dropped rows.

SenthamizhanV replied:
This was just to get a unique set of (state, district) pairs in order to map the District LGD codes. I ensured no information is lost here.

beets replied:
ok thanks

beets commented on scripts/india_wris/base.py:
```diff
@@ -0,0 +1,156 @@
+                                how='left')
+        self.df['dcid'] = self.df.apply(lambda x: ''.join([
+            'india_wris/',
+            str(x['DistrictCode']), '_', ''.join(
```
Are Station Names guaranteed to be unique within districts, or is there another identifier that can be added to guarantee uniqueness in these generated ID's.

SenthamizhanV replied:
Yes, the station names are unique. I confirmed this with the lat long. All stations had unique lat long values.

beets replied:
sounds good.

beets commented on scripts/india_wris/util/groundWater.json:
```diff
@@ -0,0 +1,122 @@
+[
```
Can you add documentation for the json format / properties in the README

SenthamizhanV replied:
Added documentation for json files under respective README files.

beets commented on scripts/india_wris/util/groundWater.json:
```diff
@@ -0,0 +1,122 @@
+            {
+                "name": "Arsenic (As)",
+                "statvar": "Arsenic",
+                "unit": "dcid:MicrogramsPerLitre"
```
Nit: seems inconsistent to add the dcid prefix for unit, but not the statvar. Choose one -- it's probably easiest to omit it here, and add it programatically if the unit is added.

SenthamizhanV replied:
I implemented this.

beets commented on scripts/india_wris/util/groundWater.json:
```diff
@@ -0,0 +1,122 @@
+                "unit": "dcid:MicrogramsPerLitre"
+            },
+            {
+                "name": "carbonate (CO3)",
```
Is this lowercase by accident (or is that how it is structured in csv)

beets reviewed (commented):
Please run "git rebase" -- that will remove the diffs from other PR's that have been merged since this branch was created.

https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase

Probably this command should work for you as well, depending on your configuration: `git rebase -Xours upstream/master`

SenthamizhanV commented on scripts/README.md:
```diff
@@ -1,9 +0,0 @@
-# Data Commons Scripts
```
Added back `scripts/README.md` and `scripts/__init__.py`

SenthamizhanV commented on scripts/india/geo/districts.py:
```diff
@@ -59,6 +59,12 @@ def get_district_name_to_lgd_code_mapping(self, state_name, district_name):
         if len(df_districts.index) == 1:
             return list(df_districts["LGDDistrictCode"])[0]
+        
+        # Try match using closestDistrictLabel (Close Name Alternatives)
```
Added a [new test](https://github.com/datacommonsorg/data/blob/9e1ab51d6e1551ff631a7f65492e92af9f7d8c26/scripts/india/geo/districts_test.py#L47)

SenthamizhanV commented on scripts/india_wris/India_WRIS_Ground/India_WRIS_Ground.tmcf:
```diff
@@ -0,0 +1,232 @@
+observationPeriod: "P1Y"
+variableMeasured: dcid:WaterQuality_Concentration_GroundWater_Arsenic
+measuredProperty: dcs:concentration
+value: C:India_WRIS_Ground->Arsenic (As)
```
Yes, it is the column name in the dataset. I have added quotes around them as you suggested.

SenthamizhanV commented on scripts/india_wris/India_WRIS_Rainfall/README.md:
```diff
@@ -0,0 +1,23 @@
+
+Source link: [https://indiawris.gov.in/wris/#/DataDownload](https://indiawris.gov.in/wris/#/DataDownload)
+
+
```
In India_WRIS_Rainfall, we don't have any scripts. Since the data is small and there are only three statvars, I did the preprocessing manually. But I added commands to run the code in other two directories: [India_WRIS_Surface](https://github.com/datacommonsorg/data/blob/9e1ab51d6e1551ff631a7f65492e92af9f7d8c26/scripts/india_wris/India_WRIS_Surface/README.md) and [India_WRIS_Ground](https://github.com/datacommonsorg/data/blob/9e1ab51d6e1551ff631a7f65492e92af9f7d8c26/scripts/india_wris/India_WRIS_Ground/README.md).

beets replied:
could you please add the detailed steps you did to the required data? e.g. int the link to download data, there are some options (which application / report), then what were the manual steps you did to that data? imagine someone else needs to come a year from now to refresh the data, what would they need to know?

SenthamizhanV commented:
@beets I tried rectifying lint errors multiple times but the checks keep failing. Would you please tell me which files are throwing the error? I don't have access to the build logs.

beets commented:
hm, the error seems to be in an unrelated file.. and the fix shows what's in the repo (below). it seems the version of the file in your commits is [fairly old (from oct 14)](https://github.com/datacommonsorg/data/blob/5b78901f0a146be549dee0358f28e5b057848106/scripts/google_research/open-covid-19/generate_csv_mcf.py) while what's at [HEAD is more recent](https://github.com/datacommonsorg/data/blob/master/scripts/google_research/open-covid-19/generate_csv_mcf.py)

try doing `git pull upstream master`. there are also some conflicts to be resolved. 

```
Step #2 - "python_format_check": --- scripts/google_research/open-covid-19/generate_csv_mcf.py	(original)
Step #2 - "python_format_check": +++ scripts/google_research/open-covid-19/generate_csv_mcf.py	(reformatted)
Step #2 - "python_format_check": @@ -425,7 +425,9 @@
Step #2 - "python_format_check":      """Generates node MCFs for geos that do not come with DCIDs
Step #2 - "python_format_check":      for local resolution."""
Step #2 - "python_format_check":      mcfs = []
Step #2 - "python_format_check": -    template = ('Node: {iso}\n' 'typeOf: schema:Place\n' 'isoCode: "{iso}"\n')
Step #2 - "python_format_check": +    template = ('Node: {iso}\n'
Step #2 - "python_format_check": +                'typeOf: schema:Place\n'
Step #2 - "python_format_check": +                'isoCode: "{iso}"\n')
Step #2 - "python_format_check":      for value in observation_abouts.unique():
Step #2 - "python_format_check":          if value.startswith('l:'):
Step #2 - "python_format_check":              mcfs.append(template.format_map({'iso': value[2:]}))
Step #2 - "python_format_check": ERROR: Fix lint errors by running ./run_test.sh -f
```

SenthamizhanV commented:
Thank you @beets. I pulled upstream/master and it resolved the errors. All checks are passed now.

beets commented on scripts/india_wris/India_WRIS_Surface/India_WRIS_Surface.mcf:
```diff
@@ -0,0 +1,235 @@
+contaminatedThing: dcs:SurfaceWater
+contaminant: dcs:Aluminium
+measuredProperty: dcs:concentration
+measurementMethod: WRIS_India
```
I don't see the update to the mf?

beets commented on scripts/india_wris/India_WRIS_Surface/India_WRIS_Surface.mcf:
```diff
@@ -0,0 +1,235 @@
+contaminant: dcs:Aluminium
+measuredProperty: dcs:concentration
+measurementMethod: WRIS_India
+statType: measuredValue
```
I don't see hte update to the mcf?

beets reviewed (commented):
thanks for the updates! a few further comments.

SenthamizhanV commented on city_not_found.txt:
```diff
@@ -0,0 +1 @@
+ak, bristol bay borough, 852.0
```
I checked this and it kept appearing when I ran scripts/ level python tests. Then I checked the files and this file is being created from the [FBI Crime import](https://github.com/datacommonsorg/data/blob/6a59e665c2d635f348cd2dcd7d798ec4a5978312/scripts/fbi/crime/preprocess.py#L278). That [PR](https://github.com/datacommonsorg/data/commit/4753c6eb5f7b702d0a0012f7a3170ad6da49e70c#diff-352e3b86c933f0b5170dd1c1f40a6fd3bbd6dcd49ed3c08be81dcad693888c52) was created and pushed by Hanlu Chen.

SenthamizhanV commented:
@beets I have implemented the changes you suggested. Also added an exception clause in [base.py](https://github.com/datacommonsorg/data/pull/549/files#diff-5b2d851d959e09a4ec9834bdd57973d0cd3a7a4d7490ff83f7cd59d019917fc9) to catch district to lgd code mapping errors.

Also, there are some files from US census in this latest commit. Those were created or modified when I ran scripts/ level python tests.

beets commented:
Thanks for the updates!

Regarding the extra files -- we should fix up the other tests so they don't output files (or we ignore them in git). Before we fix those, could you remove them from this PR? After running the tests, don't run `git add .` but rather add only files that are relevant for merging. I find it's easier to do this from [github desktop](https://desktop.github.com/)

SenthamizhanV commented:
Thanks for the fix in the [new PR](https://github.com/datacommonsorg/data/pull/613), @beets. I will use that after the PR's been merged.

beets commented:
I just merged the PR -- please try. You might need to manually revert the extra added files from the PR even after they've been added to gitignore.

SenthamizhanV commented:
@beets So the modified US census files in my previous commits are now deleted in datacommons master branch. So I have deleted those files from this PR, too. There are other files showing up in Files Changed now. But they appeared after I pulled the datacommons master branch today.

SenthamizhanV commented:
@beets I changed the StatVar names in this import to make it as similar to Superfund import. My latest commit fails the lint checks. But, my local tests are running successfully (screenshot attached). Can you please check the logs and share the files that are failing the lint checks?

![lint](https://user-images.githubusercontent.com/30504842/161228926-1248c806-d2d1-4be2-b80e-d485af2f50ee.png)

beets commented:
Hm, it seems to be lint errors. Do you have local changes that haven't been pushed?

```bash
Already have image: python:3.7
### Testing lint
Installing Python requirements
#### Testing Python lint
--- scripts/google_research/open-covid-19/generate_csv_mcf.py	(original)
+++ scripts/google_research/open-covid-19/generate_csv_mcf.py	(reformatted)
@@ -425,7 +425,9 @@
     """Generates node MCFs for geos that do not come with DCIDs
     for local resolution."""
     mcfs = []
-    template = ('Node: {iso}\n' 'typeOf: schema:Place\n' 'isoCode: "{iso}"\n')
+    template = ('Node: {iso}\n'
+                'typeOf: schema:Place\n'
+                'isoCode: "{iso}"\n')
     for value in observation_abouts.unique():
         if value.startswith('l:'):
             mcfs.append(template.format_map({'iso': value[2:]}))
ERROR: Fix lint errors by running ./run_test.sh -f
```

SenthamizhanV commented:
That one file raises lint issues whenever I pull upstream. Not sure why. I just copied the code of that file from master branch and pasted locally and pushed it. The issue is resolved now. Thank you @beets!
